# Day 2: Dive!

(From [Advent of Code 2021, day 2](https://adventofcode.com/2021/day/2))

In [1]:
# Libraries used:
import pandas as pd
import numpy as np

## Part 1

Now, you need to figure out how to pilot this thing.

It seems like the submarine can take a series of commands like `forward 1`, `down 2`, or `up 3`:

 - `forward X` increases the horizontal position by X units;
 - `down X` increases the depth by X units; and
 - `up X` decreases the depth by X units.

 Note that since you're on a submarine, `down` and `up` affect your depth, and so they have the opposite result of what you might expect.

The submarine seems to already have a planned course (your puzzle input). You should probably figure out where it's going. For example:

```txt
forward 5
down 5
forward 8
up 3
down 8
forward 2
```

Your horizontal position and depth both start at 0. The steps above would then modify them as follows:

 - `forward 5` adds 5 to your horizontal position, a total of 5.
 - `down 5` adds 5 to your depth, resulting in a value of 5.
 - `forward 8` adds 8 to your horizontal position, a total of 13.
 - `up 3` decreases your depth by 3, resulting in a value of 2.
 - `down 8` adds 8 to your depth, resulting in a value of 10.
 - `forward 2` adds 2 to your horizontal position, a total of 15.

After following these instructions, you would have a horizontal position of 15 and a depth of 10. (Multiplying these together produces 150.)

Calculate the horizontal position and depth you would have after following the planned course. *What do you get if you multiply your final horizontal position by your final depth?*


### Solution

In [2]:
# Read in input data
## Break down submarine commands 
## into direction (forward, down, up) and distance (value)
data = pd.read_csv('input.txt', header=None, names=['direction', 'value'], delimiter=" ")

# Quick look at data
data.head()

direction  value
0   forward      9
1   forward      7
2      down      7
3      down      3
4   forward      2

In [3]:
# Calculate horizontal position
horizontal_pos = sum(data[data['direction']=='forward']['value'])

# Calculate vertical position (depth)
depth_increase = sum(data[data['direction']=='down']['value'])
depth_decrease = sum(data[data['direction']=='up']['value'])
depth = depth_increase - depth_decrease

Multiplying horizontal position and depth produces:

In [4]:
horizontal_pos * depth

1882980

*Answer*: 1882980

## Part 2

Based on your calculations, the planned course doesn't seem to make any sense. You find the submarine manual and discover that the process is actually slightly more complicated.

In addition to horizontal position and depth, you'll also need to track a third value, **aim**, which also starts at `0`. The commands also mean something entirely different than you first thought:
 - `down X` increases your aim by X units;
 - `up X` decreases your aim by X units; and
 - `forward X` does two things:
    - it increases your horizontal position by X units; and
    - it increases your depth by your aim multiplied by X.
    
Again note that since you're on a submarine, `down` and `up` do the opposite of what you might expect: "down" means aiming in the positive direction.

Now, the above example does something different:

 - `forward 5` adds 5 to your horizontal position, a total of 5. Because your aim is 0, your depth does not change.
 - `down 5` adds 5 to your aim, resulting in a value of 5.
 - `forward 8` adds 8 to your horizontal position, a total of 13. Because your aim is 5, your depth increases by 8*5=40.
 - `up 3` decreases your aim by 3, resulting in a value of 2.
 - `down 8` adds 8 to your aim, resulting in a value of 10.
 - `forward 2` adds 2 to your horizontal position, a total of 15. Because your aim is 10, your depth increases by 2*10=20 to a total of 60.

After following these new instructions, you would have a horizontal position of 15 and a depth of 60. (Multiplying these produces 900.)

Using this new interpretation of the commands, calculate the horizontal position and depth you would have after following the planned course. *What do you get if you multiply your final horizontal position by your final depth?*

### Solution

In [5]:
# Calculate aim for every command instance
def get_aim(data):
    if data['direction'] == 'forward':
        return 0
    elif data['direction'] == 'up':
        return 0 - data['value']
    else:
        return data['value']

data['aim'] = data.apply(get_aim, axis=1)

# Peak at data
data.head()

direction  value  aim
0   forward      9    0
1   forward      7    0
2      down      7    7
3      down      3    3
4   forward      2    0

In [6]:
# Calculate "total" aim at each command instance
## Transform aim values into cumulative sum
## at every command instance
data['aim'] = data['aim'].cumsum()

# Peak at data
data.head(20)

direction  value  aim
0    forward      9    0
1    forward      7    0
2       down      7    7
3       down      3   10
4    forward      2   10
5    forward      3   10
6    forward      7   10
7       down      6   16
8    forward      7   16
9       down      7   23
10   forward      9   23
11      down      9   32
12        up      2   30
13      down      5   35
14        up      1   34
15   forward      5   34
16   forward      6   34
17        up      4   30
18      down      3   33
19      down      4   37

In [7]:
# Calculate depth after taking aim into account
## Note that in our updated case,
## only "forward" commands directly effect
## the depth of the submarine
forward_commands = data[data['direction']=='forward']
depth = sum(forward_commands['value'] * forward_commands['aim'])

Multiplying final horizontal position and final depth produces:

In [8]:
horizontal_pos * depth

1971232560

*Answer*: 1971232560